### PR TITLE
Support for Kubernetes via Terraform and Manifests

### PR DESCRIPTION
**Pull request by LumosViridi**
# Support for Kubernetes via Terraform and Manifests

Adding basic support for the following Kubernetes resources

- persistent volume
  - server
  - database
- persistent volume claim
  - server
  - database
- deployment
  - server
  - database
- ingress
  - server
- service
  - server
  - database
- secret
  - server

### CODE REVIEW

File: packages/twenty-docker/k8s/terraform/variables.tf [hunk 1]
```diff
@@ -0,0 +1,25 @@
+### TwentyCRM
```
[the remark this answers is not in the document]
Comment (LumosViridi) [reply packages/twenty-docker/k8s/terraform/variables.tf]: Removed line

File: packages/twenty-docker/k8s/terraform/deployment-server.tf [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+          tty   = true
+
+          security_context {
+            allow_privilege_escalation = true
+            privileged                 = true
+            run_as_user                = 1000
+          }
+
+          env {
+            name  = "PORT"
```
[the remark this answers is not in the document]
Comment (LumosViridi) [reply packages/twenty-docker/k8s/terraform/deployment-server.tf:47]: Defaulting to false

File: packages/twenty-docker/k8s/terraform/deployment-db.tf [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        labels = {
+          app = "${local.twentycrm_app_name}-db"
+        }
+      }
+
+      spec {
+        # security_context {
+        #   fs_group = 0
+        # }
+        container {
```
[the remark this answers is not in the document]
Comment (LumosViridi) [reply packages/twenty-docker/k8s/terraform/deployment-db.tf:37]: This is a twentyCRM default and can be updated as needed

File: packages/twenty-docker/k8s/README.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
+# README
+
+## Overview
+
+This repository contains Kubernetes manifests and Terraform files to help you deploy and manage the TwentyCRM application. The files are located in the `packages/twenty-docker/k8s` directory.
+
+## Prerequisites
+
+Before using these files, ensure you have the following installed and configured on your system:
```
[the remark this answers is not in the document]
Comment (LumosViridi) [reply packages/twenty-docker/k8s/README.md:9]: This is a folder path, not a URL, Leaving for accuracy.

**Comment by FelixMalfait** [comment]
Oh nice thanks :). I've never done any Terraform so I'm not able to review this but I know some people will be happy to see this

**Comment by sysaeon** [comment]
Thanks for adding this! It motivated me to get Twenty setup in my k3s cluster for testing.

I ended up using the bjw-s App Template Helm Chart with many of the values you provided here in the manifests. Also, using Flux instead of Terraform.

Had to define FRONT_BASE_URL due to this error:

`- property FRONT_BASE_URL has failed the following constraints: isUrl`

Also, the service for twenty-postgres shows up as `twenty-twenty-postgres`, which changed the URL (for me) to:

`postgres://twenty:twenty@twenty-twenty-postgres.twentycrm.svc.cluster.local/default`

Does anyone know if it's possible to change the postgres password without building a new docker image? It would make sense to put the postgres password and URL into secrets and call them with something like this (under each deployment):

```
- name: POSTGRES_PASSWORD
  valueFrom:
    secretKeyRef:
      name: twenty-config-secrets
      key: pgPassword
---
- name: PG_DATABASE_URL
  valueFrom:
    secretKeyRef:
      name: twenty-config-secrets
      key: pgDatabaseUrl
```

Otherwise, everything appears to be running well.

**Comment by LumosViridi** [comment]
> Thanks for adding this! It motivated me to get Twenty setup in my k3s cluster for testing.
> 
> I ended up using the bjw-s App Template Helm Chart with many of the values you provided here in the manifests. Also, using Flux instead of Terraform.
> 
> Had to define FRONT_BASE_URL due to this error:
> 
> `- property FRONT_BASE_URL has failed the following constraints: isUrl`
> 
> Also, the service for twenty-postgres shows up as `twenty-twenty-postgres`, which changed the URL (for me) to:
> 
> `postgres://twenty:twenty@twenty-twenty-postgres.twentycrm.svc.cluster.local/default`
> 
> Does anyone know if it's possible to change the postgres password without building a new docker image? It would make sense to put the postgres password and URL into secrets and call them with something like this (under each deployment):
> 
> ```
> - name: POSTGRES_PASSWORD
>   valueFrom:
>     secretKeyRef:
>       name: twenty-config-secrets
>       key: pgPassword
> ---
> - name: PG_DATABASE_URL
>   valueFrom:
>     secretKeyRef:
>       name: twenty-config-secrets
>       key: pgDatabaseUrl
> ```
> 
> Otherwise, everything appears to be running well.

Thanks for your comment @sysaeon!

The pg password caused a lot of trouble for me as well. I spent a few hours banging my head on my keyboard trying to get any value other than `twenty` to work. I was only able to establish a connection with the hardcoded / default value. 🫤

I had issues with the `FRONT_BASE_URL` as well but it looks like you were able to work through it.

**Comment by sashahilton00** [comment]
I'd just add, the use of `ReadWriteMany` for the persistent volumes will cause problems with most cloud providers. It may work in a k3s setup, but iirc GCP, AWS and DO do not support it for their storage classes.

**Comment by LumosViridi** [comment]
> I'd just add, the use of `ReadWriteMany` for the persistent volumes will cause problems with most cloud providers. It may work in a k3s setup, but iirc GCP, AWS and DO do not support it for their storage classes.

Good call! Most default cases like AWS EBS-CSI don't support `ReadWriteMany` but AWS EFS-CSI now has native support.

I tried to convey in the README that storage is highly opinionated to the users environment and should be expected to be modified. 

I update this to `ReadWriteOnce` since it will work in more use-cases than `ReadWriteMany`. Hopefully it helps others use this more easily! 🚀

**Comment by FelixMalfait** [comment]
Hey sorry we left this open for so long. 
We are still focused on product development/stability and getting out of alpha/beta asap so deployment isn't a core priority yet, but hopefully it will be once things become more stable. Even though we didn't do a proper review, I'm adding this to the doc and adding this as a "community-led effort" as I'm sure it can be very helpful to many! Thanks a lot!

**Pull request merged**